File: pulsar-skeleton/src/main/java/ai/platon/pulsar/common/options/PulsarOptions.java
package ai.platon.pulsar.common.options;

import ai.platon.pulsar.common.StringUtil;
import ai.platon.pulsar.common.config.Parameterized;
import com.beust.jcommander.JCommander;
import com.beust.jcommander.ParameterException;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Created by vincent on 17-4-12.
 * Copyright @ 2013-2017 Platon AI. All rights reserved
 */
public class PulsarOptions implements Parameterized {
    public static final Logger LOG = LoggerFactory.getLogger(PulsarOptions.class);

    public static final String DEFAULT_DELIMETER = " ";

    protected boolean expandAtSign = true;
    protected String args = "";
    protected String[] argv;
    protected Set<Object> objects = new HashSet<>();
    protected JCommander jc;

    public PulsarOptions() {
        this.argv = new String[]{};
    }

    public PulsarOptions(String args) {
        this(split(args));
        this.args = args;
    }

    public PulsarOptions(String[] argv) {
        this.argv = argv;
        for (int i = 0; i < this.argv.length; ++i) {
            // Since space can not appear in dynamic parameters in command line, we use % instead
            this.argv[i] = this.argv[i].replaceAll("%", " ");
        }
        if (args.isEmpty()) {
            args = StringUtils.join(argv);
        }
    }

    public PulsarOptions(Map<String, String> argv) {
        this(argv.entrySet().stream()
                .map(e -> e.getKey() + DEFAULT_DELIMETER + e.getValue())
                .collect(Collectors.joining(DEFAULT_DELIMETER)));
    }

    public String getArgs() {
        return args;
    }

    @Nonnull
    public static String normalize(String args) {
        return normalize(args, ",");
    }

    @Nonnull
    public static String normalize(String args, String seps) {
        return StringUtils.replaceChars(args, seps, StringUtils.repeat(' ', seps.length()));
    }

    @Nonnull
    public static String[] split(String args) {
        return normalize(args).split("\\s+");
    }

    public boolean getExpandAtSign() {
        return expandAtSign;
    }

    public void setExpandAtSign(boolean expandAtSign) {
        this.expandAtSign = expandAtSign;
    }

    public void setObjects(Object... objects) {
        this.objects.clear();
        this.objects.addAll(Lists.newArrayList(objects));
    }

    public void addObjects(Object... objects) {
        this.objects.addAll(Lists.newArrayList(objects));
    }

    public boolean parse() {
        try {
            doParse();
            return true;
        } catch (Throwable e) {
            LOG.warn(StringUtil.stringifyException(e));
        }

        return false;
    }

    public void parseOrExit() {
        parseOrExit(Sets.newHashSet());
    }

    protected void parseOrExit(Set<Object> objects) {
        try {
            addObjects(objects);
            doParse();

            if (isHelp()) {
                jc.usage();
                System.exit(0);
            }
        } catch (ParameterException e) {
            System.out.println(e.toString());
            System.exit(0);
        }
    }

    private void doParse() {
        if (!objects.contains(this)) {
            objects.add(this);
        }

        if (jc == null) {
            jc = new JCommander(objects);
        }

        jc.setAcceptUnknownOptions(true);
        jc.setAllowParameterOverwriting(true);
//      jc.setAllowAbbreviatedOptions(false);
        jc.setExpandAtSign(expandAtSign);
        // log.debug(StringUtils.join(argv, " "));
        if (argv != null) {
            jc.parse(argv);
        }
    }

    public boolean isHelp() {
        return false;
    }

    public void usage() {
        jc.usage();
    }

    public String toCmdLine() {
        return getParams().withKVDelimiter(" ").formatAsLine().replaceAll("\\s+", " ");
    }

    public String[] toArgs() {
        return getParams().withKVDelimiter(" ").formatAsLine().replaceAll("\\s+", " ").split("\\s+");
    }

    @Override
    public boolean equals(Object other) {
        return (other instanceof PulsarOptions) && this.toString().equals(other.toString());
    }

    @Override
    public String toString() {
        return StringUtils.join(argv, " ");
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/common/options/PulsarOptions.kt
package ai.platon.pulsar.common.options

import ai.platon.pulsar.common.config.Parameterized
import com.beust.jcommander.JCommander
import com.beust.jcommander.ParameterException
import org.apache.commons.lang3.StringUtils
import org.slf4j.LoggerFactory
import java.util.*
import java.util.regex.Pattern
import kotlin.system.exitProcess

/**
 * Created by vincent on 17-4-12.
 * Copyright @ 2013-2017 Platon AI. All rights reserved
 */
open class PulsarOptions(
    /**
     * The argument vector
     * */
    val argv: Array<String>
) : Parameterized {
    protected val logger = LoggerFactory.getLogger(PulsarOptions::class.java)

    init { normalize(argv) }

    var expandAtSign = true
    var acceptUnknownOptions = true
    var allowParameterOverwriting = true
    // arguments
    val args: String get() = argv.joinToString(DEFAULT_DELIMETER)
    private val registeredObjects: MutableSet<Any> = HashSet()
    protected lateinit var jc: JCommander

    open var isHelp: Boolean = false

    init { addObjects(this) }

    constructor(): this(arrayOf())

    constructor(args: String): this(split(args.trim()))

    constructor(argv: Map<String, String>)
            : this(argv.entries.joinToString(DEFAULT_DELIMETER) { it.key + DEFAULT_DELIMETER + it.value })

    fun setObjects(vararg objects: Any) {
        this.registeredObjects.clear()
        addObjects(objects)
    }

    fun addObjects(vararg objects: Any) {
        objects.toCollection(this.registeredObjects)
    }

    open fun parse(): Boolean {
        try {
            doParse()
        } catch (e: Throwable) {
            logger.warn("Failed to parse \n$args", e)
            return false
        }

        return true
    }

    open fun parseOrExit() {
        parseOrExit(mutableSetOf())
    }

    protected open fun parseOrExit(objects: Set<Any>) {
        try {
            addObjects(objects)
            doParse()

            if (isHelp) {
                jc.usage()
                exitProcess(0)
            }
        } catch (e: ParameterException) {
            println(e.toString())
            exitProcess(0)
        }
    }

    /**
     * TODO: fix bug to handle overwriting boolean field with arity = 0, e.g. "-parse -parse"
     * */
    private fun doParse() {
        registeredObjects.add(this)

        jc = JCommander.newBuilder()
                .acceptUnknownOptions(acceptUnknownOptions)
                .allowParameterOverwriting(allowParameterOverwriting)
                .expandAtSign(expandAtSign).build()
        registeredObjects.forEach { jc.addObject(it) }

        if (argv.isNotEmpty()) {
            jc.parse(*argv)
        }
    }

    open fun usage() {
        jc.usage()
    }

    open fun toCmdLine(): String {
        return params.withKVDelimiter(" ").formatAsLine()
                .replace("\\s+".toRegex(), " ")
    }

    open fun toArgsMap(): Map<String, String> {
        return params.asStringMap()
    }

    open fun toMutableArgsMap(): MutableMap<String, String> {
        return params.asStringMap()
    }

    open fun toArgv(): Array<String> {
        return params.withKVDelimiter(" ").formatAsLine()
                .split("\\s+".toRegex())
                .toTypedArray()
    }

    override fun hashCode(): Int {
        return args.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return other is PulsarOptions && args == other.args
    }

    override fun toString(): String {
        return args
    }

    companion object {
        const val DEFAULT_DELIMETER = " "
        val CMD_SPLIT_PATTERN = Pattern.compile("\"[^\"\\\\]*(?:\\\\.[^\"\\\\]*)*\"|\\S+")

        /**
         * Normalize the raw arguments, convert old version args to current version
         * */
        @JvmOverloads
        fun normalize(args: String, seps: String = ","): String {
            var args1 = StringUtils.replaceChars(args, seps, StringUtils.repeat(' ', seps.length))
            // in old version, -cacheContent has arity 0, but current version is 1, we need a convert
            args1 = OptionUtils.arity0ToArity1(args1, "-cacheContent")
            args1 = OptionUtils.arity0ToArity1(args1, "-storeContent")

            return args1
        }

        /**
         * Since space can not appear in dynamic parameters in command line, we use % instead.
         * */
        fun normalize(argv: Array<String>) {
            for (i in argv.indices) {
                argv[i] = argv[i]
                    .replace("%", " ")
                    .replace("%20", " ")
            }
        }

        /**
         * Split a command line into argument vector (argv).
         * @see {https://stackoverflow.com/questions/36292591/splitting-a-nested-string-keeping-quotation-marks/36292778}
         */
        fun split(args: String): Array<String> {
            val matcher = CMD_SPLIT_PATTERN.matcher(normalize(args))
            matcher.reset()
            val result = ArrayList<String>()
            while (matcher.find()) {
                result.add(matcher.group(0))
            }
            return result.toTypedArray()
        }
    }
}
